File: backend/src/voice_clone/schemas/voice_clone.py
"""Voice clone schemas."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, HttpUrl

from voice_clone.models.writing_sample import SourceType


class VoiceCloneCreate(BaseModel):
    """Schema for creating a voice clone."""

    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    tags: list[str] = Field(default_factory=list)


class VoiceCloneUpdate(BaseModel):
    """Schema for updating a voice clone."""

    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    tags: list[str] | None = None


class WritingSampleCreate(BaseModel):
    """Schema for creating a writing sample."""

    source_type: SourceType
    content: str | None = None  # For paste
    source_url: HttpUrl | None = None  # For URL
    # File uploads handled separately via multipart form


class WritingSampleResponse(BaseModel):
    """Schema for writing sample response."""

    id: str
    voice_clone_id: str
    source_type: SourceType
    content: str
    source_url: str | None
    original_filename: str | None
    word_count: int
    created_at: datetime

    model_config = {"from_attributes": True}


class VoiceDnaResponse(BaseModel):
    """Schema for Voice DNA response."""

    id: str
    voice_clone_id: str
    version: int
    dna_data: dict[str, Any]
    analysis_metadata: dict[str, Any] | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ConfidenceScoreResponse(BaseModel):
    """Schema for confidence score response."""

    score: int = Field(..., ge=0, le=100)
    breakdown: dict[str, int]


class VoiceCloneResponse(BaseModel):
    """Schema for voice clone response."""

    id: str
    user_id: str
    name: str
    description: str | None
    tags: list[str]
    is_merged: bool
    confidence_score: int
    current_dna_id: str | None
    created_at: datetime
    updated_at: datetime | None
    sample_count: int = 0
    total_word_count: int = 0

    model_config = {"from_attributes": True}


class VoiceCloneListResponse(BaseModel):
    """Schema for paginated voice clone list response."""

    items: list[VoiceCloneResponse]
    total: int
    page: int
    per_page: int
    pages: int


class VoiceCloneDetailResponse(VoiceCloneResponse):
    """Schema for detailed voice clone response including samples and DNA."""

    samples: list[WritingSampleResponse] = Field(default_factory=list)
    current_dna: VoiceDnaResponse | None = None


class MergeSourceConfig(BaseModel):
    """Schema for merge source configuration."""

    voice_clone_id: str
    element_weights: dict[str, int] = Field(
        ...,
        description="Weights for each DNA element (0-100)",
    )


class MergeCloneRequest(BaseModel):
    """Schema for merge clone request."""

    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    sources: list[MergeSourceConfig] = Field(
        ...,
        min_length=2,
        max_length=5,
        description="2-5 source voice clones with element weights",
    )

File: frontend/src/types/voice-clone.ts
/**
 * Voice clone type definitions
 */

export interface VoiceClone {
  id: string
  user_id: string
  name: string
  description?: string
  tags: string[]
  is_merged: boolean
  confidence_score: number
  current_dna_id?: string
  created_at: string
  updated_at: string
}

export interface VoiceCloneCreate {
  name: string
  description?: string
  tags?: string[]
}

export interface VoiceCloneUpdate {
  name?: string
  description?: string
  tags?: string[]
}

export interface VoiceCloneListResponse {
  items: VoiceClone[]
  total: number
  page: number
  per_page: number
  pages: number
}

export interface VoiceCloneDetailResponse extends VoiceClone {
  samples: WritingSample[]
  current_dna?: VoiceDna
}

export enum SourceType {
  PASTE = "paste",
  FILE = "file",
  URL = "url",
}

export interface WritingSample {
  id: string
  voice_clone_id: string
  source_type: SourceType
  content: string
  source_url?: string
  original_filename?: string
  word_count: number
  created_at: string
}

export interface WritingSampleCreate {
  source_type: SourceType
  content?: string
  source_url?: string
  file?: File
}

export interface VoiceDna {
  id: string
  voice_clone_id: string
  version: number
  dna_data: VoiceDnaData
  analysis_metadata?: Record<string, unknown>
  created_at: string
}

export interface VoiceDnaData {
  vocabulary_patterns: Record<string, unknown>
  sentence_structure: Record<string, unknown>
  paragraph_structure: Record<string, unknown>
  tone_markers: Record<string, unknown>
  rhetorical_devices: Record<string, unknown>
  punctuation_habits: Record<string, unknown>
  opening_patterns: Record<string, unknown>
  closing_patterns: Record<string, unknown>
  humor_and_personality: Record<string, unknown>
  distinctive_signatures: Record<string, unknown>
}

export interface ConfidenceScore {
  score: number
  breakdown: {
    word_count_score: number
    sample_count_score: number
    content_variety_score: number
    length_distribution_score: number
    consistency_score: number
  }
}

export interface MergeSourceConfig {
  voice_clone_id: string
  element_weights: Record<string, number>
}

export interface MergeCloneRequest {
  name: string
  description?: string
  tags?: string[]
  sources: MergeSourceConfig[]
}
